In [2]:
import os
import json
import numpy as np
import pandas as pd
import xgboost as xgb
from scipy.stats import spearmanr

folder_path = r"D:\OneDrive\Trading\Market Making\data\runs\run_20260905_081614"

fills = pd.read_parquet(os.path.join(folder_path, "fills.parquet"))

fills

,ts,exchange_latency,time_to_fill,symbol,side,price,price_tick,qty,cum_qty,is_maker,...,ask_distance_spread,snap_mid,signed_markout_100ms,signed_adverse_100ms,signed_markout_500ms,signed_adverse_500ms,signed_markout_1000ms,signed_adverse_1000ms,signed_markout_5000ms,signed_adverse_5000ms
0,1788596208003,29016,28977,PEPEUSDT,SELL,0.000004,352,312945.0,312945.0,True,...,1.000000e-08,0.000004,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09
1,1788596208102,41,0,PEPEUSDT,SELL,0.000004,352,312945.0,312945.0,False,...,0.000000e+00,0.000004,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09
2,1788596622379,248909,248868,PEPEUSDT,BUY,0.000004,351,830286.0,830286.0,True,...,1.000000e-08,0.000004,5.000000e-09,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09,-5.000000e-09,5.000000e-09,-5.000000e-09


PS D:\OneDrive\Trading\Market Making> ./mm.exe
D://OneDrive//Trading//Market Making//manifest.json
exchange: binance_spot, mode: live, instrument: pepeusdt, instrument_upper: PEPEUSDT
exchange_latency: 0, gamma: 0.1, base_size: 1.1, max_inv: 3
initial_cash: 100
folder_path: data/runs/run_20260830_045447, events_chunk: 50, snapshots_chunk: 50
trades_chunk: 15, quotes_chunk: 10, fills_chunk: 1
host: 0.0.0.0, port: 8000
hostname: stream.binance.com, base_url: api.binance.com, endpoint: api/v3
tick_size: 1e-08, price_precision: 8
step_size: 1, qty_precision: 0
min_qty: 1, max_qty: 9.22337e+10
min_notional_USDT: 1, apply_min_to_market: true
INITIALIZED: regime_model_pepe target: detect_regime
events_file_path: data/runs/run_20260905_081614/events_000000.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000000.parquet
trades_file_path: data/runs/run_20260905_081614/trades_000000.parquet
quotes_file_path: data/runs/run_20260905_081614/quotes_000000.parquet
fills_file_path: data/runs/run_20260905_081614/fills_000000.parquet
[HTTP] broker connected: api.binance.com

===== HTTP RESPONSE =====
maker_fee_rate: 0.001 (0.1%)
taker_fee_rate: 0.001 (0.1%)
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 76 ms
LIVE SPOT SOCKETS STARTED
WS RUNNING ON http://localhost:8000
DASHBOARD CLIENT CONNECTED
SNAPSHOT FETCHED: 8774088972
ORDER BOOK INITIALIZED
BUFFER U: 8774088973 snapshot_id + 1: 8774088973 u: 8774088973
BOOK SYNCHRONIZED
LIVE BOOK RUNNING
[USER STREAM] resolving ws-api.binance.com
[USER STREAM] subscription response: {"id":"mm-core","status":200,"result":{"subscriptionId":0},"rateLimits":[{"rateLimitType":"REQUEST_WEIGHT","interval":"MINUTE","intervalNum":1,"limit":6000,"count":99}]}
[USER STREAM] subscribed, subscriptionId=0
Starting execution loop
- PLACE LIMIT ORDER - client_oid: MM-8da57fc5657843b0, status: PENDING_NEW, timestamp: 1788596178548
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596178946,"s":"PEPEUSDT","c":"MM-8da57fc5657843b0","S":"BUY","o":"LIMIT","f":"GTC","q":"312945.00","p":"0.00000351","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"NEW","X":"NEW","r":"NONE","i":5410202117,"l":"0.00","z":"0.00","L":"0.00000000","n":"0","N":null,"T":1788596178946,"t":-1,"I":11553801302,"w":true,"m":false,"M":false,"O":1788596178946,"Z":"0.00000000","Y":"0.00000000","Q":"0.00000000","W":1788596178946,"V":"EXPIRE_MAKER"}}

===== HTTP RESPONSE =====
HTTP/1.1 200 OK
Content-Type: application/json;charset=UTF-8
Content-Length: 401
Connection: keep-alive
Date: Sat, 05 Sep 2026 08:16:18 GMT
Server: nginx
Vary: Accept-Encoding
x-mbx-uuid: 386c840b-94c8-4cbe-ab02-b397ea9c8d5a
x-mbx-used-weight: 100
x-mbx-used-weight-1m: 100
x-mbx-order-count-10s: 1
x-mbx-order-count-1d: 86
Strict-Transport-Security: max-age=31536000; includeSubdomains
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Content-Security-Policy: default-src 'self'
X-WebKit-CSP: default-src 'self'
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
X-Cache: Miss from cloudfront
Via: 1.1 1e6d2b7781bdded089f18a79a3020a62.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SIN2-P3
X-Amz-Cf-Id: 4Tb1oczUqwk3LN9BqY6ZOmg8-41jbH4PScOW43tjLag2mtoS-CowCA==

{"symbol":"PEPEUSDT","orderId":5410202117,"orderListId":-1,"clientOrderId":"MM-8da57fc5657843b0","transactTime":1788596178946,"price":"0.00000351","origQty":"312945.00","executedQty":"0.00","origQuoteOrderQty":"0.00000000","cummulativeQuoteQty":"0.00000000","status":"NEW","timeInForce":"GTC","type":"LIMIT","side":"BUY","workingTime":1788596178946,"fills":[],"selfTradePreventionMode":"EXPIRE_MAKER"}
- PLACE LIMIT ORDER - client_oid: MM-07bed0cb9b1c488c, status: PENDING_NEW, timestamp: 1788596178631
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596179027,"s":"PEPEUSDT","c":"MM-07bed0cb9b1c488c","S":"SELL","o":"LIMIT","f":"GTC","q":"312945.00","p":"0.00000352","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"NEW","X":"NEW","r":"NONE","i":5410202119,"l":"0.00","z":"0.00","L":"0.00000000","n":"0","N":null,"T":1788596179026,"t":-1,"I":11553801304,"w":true,"m":false,"M":false,"O":1788596179026,"Z":"0.00000000","Y":"0.00000000","Q":"0.00000000","W":1788596179026,"V":"EXPIRE_MAKER"}}

===== HTTP RESPONSE =====
HTTP/1.1 200 OK
Content-Type: application/json;charset=UTF-8
Content-Length: 402
Connection: keep-alive
Date: Sat, 05 Sep 2026 08:16:19 GMT
Server: nginx
Vary: Accept-Encoding
x-mbx-uuid: 759a9d33-72ec-4a79-82a0-810441f2cd06
x-mbx-used-weight: 100
x-mbx-used-weight-1m: 100
x-mbx-order-count-10s: 2
x-mbx-order-count-1d: 87
Strict-Transport-Security: max-age=31536000; includeSubdomains
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Content-Security-Policy: default-src 'self'
X-WebKit-CSP: default-src 'self'
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
X-Cache: Miss from cloudfront
Via: 1.1 1e6d2b7781bdded089f18a79a3020a62.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SIN2-P3
X-Amz-Cf-Id: JrYs_SYIV2axn1ZfUhQjKWSViuUPRIBgT67lQe8UaKIaa3v_BFLD3g==

{"symbol":"PEPEUSDT","orderId":5410202119,"orderListId":-1,"clientOrderId":"MM-07bed0cb9b1c488c","transactTime":1788596179026,"price":"0.00000352","origQty":"312945.00","executedQty":"0.00","origQuoteOrderQty":"0.00000000","cummulativeQuoteQty":"0.00000000","status":"NEW","timeInForce":"GTC","type":"LIMIT","side":"SELL","workingTime":1788596179026,"fills":[],"selfTradePreventionMode":"EXPIRE_MAKER"}
stream.client_oid: MM-8da57fc5657843b0, stream.exec_type: NEW
- PLACE LIMIT ORDER - client_oid: MM-8da57fc5657843b0, status: LIVE, timestamp: 1788596178946
stream NEW1
stream END
stream.client_oid: MM-07bed0cb9b1c488c, stream.exec_type: NEW
- PLACE LIMIT ORDER - client_oid: MM-07bed0cb9b1c488c, status: LIVE, timestamp: 1788596179026
stream NEW1
stream END
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 74 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 78 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 77 ms
events_file_path: data/runs/run_20260905_081614/events_000001.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000001.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 74 ms
trades_file_path: data/runs/run_20260905_081614/trades_000001.parquet
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596208005,"s":"PEPEUSDT","c":"MM-07bed0cb9b1c488c","S":"SELL","o":"LIMIT","f":"GTC","q":"312945.00","p":"0.00000352","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"TRADE","X":"FILLED","r":"NONE","i":5410202119,"l":"312945.00","z":"312945.00","L":"0.00000352","n":"0.00110157","N":"USDT","T":1788596208003,"t":560398414,"I":11553801578,"w":false,"m":true,"M":true,"O":1788596179026,"Z":"1.10156640","Y":"1.10156640","Q":"0.00000000","W":1788596179026,"V":"EXPIRE_MAKER"}}
stream.client_oid: MM-07bed0cb9b1c488c, stream.exec_type: TRADE
stream TRADE
stream FILLED
- FILLED LIMIT ORDER - client_oid: MM-07bed0cb9b1c488c, status: FILLED, timestamp: 1788596208003
stream FILLED 1
fills_file_path: data/runs/run_20260905_081614/fills_000001.parquet
stream TRADE 1
stream END
- PLACE LIMIT ORDER - client_oid: MM-0a48228afa3c42ab, status: PENDING_NEW, timestamp: 1788596207705
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596208102,"s":"PEPEUSDT","c":"MM-0a48228afa3c42ab","S":"SELL","o":"LIMIT","f":"GTC","q":"312945.00","p":"0.00000352","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"NEW","X":"NEW","r":"NONE","i":5410202348,"l":"0.00","z":"0.00","L":"0.00000000","n":"0","N":null,"T":1788596208102,"t":-1,"I":11553801787,"w":true,"m":false,"M":false,"O":1788596208102,"Z":"0.00000000","Y":"0.00000000","Q":"0.00000000","W":1788596208102,"V":"EXPIRE_MAKER"}}

===== HTTP RESPONSE =====
HTTP/1.1 200 OK{"subscriptionId":0,"event":{"e":"executionReport","E":1788596208102,"s":"PEPEUSDT","c":"MM-0a48228afa3c42ab","S":"SELL","o":"LIMIT","f":"GTC","q":"312945.00","p":"0.00000352","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"TRADE","X":"FILLED","r":"NONE","i":5410202348,"l":"312945.00","z":"312945.00","L":"0.00000352","n":"0.00110157","N":"USDT","T":1788596208102,"t":560398418,"I":11553801788,"w":false,"m":false,"M":true,"O":1788596208102,"Z":"1.10156640","Y":"1.10156640","Q":"0.00000000","W":1788596208102,"V":"EXPIRE_MAKER"}}

Content-Type: application/json;charset=UTF-8
Content-Length: 521
Connection: keep-alive
Date: Sat, 05 Sep 2026 08:16:48 GMT
Server: nginx
Vary: Accept-Encoding
x-mbx-uuid: 386c840b-94c8-4cbe-ab02-b397ea9c8d5a
x-mbx-used-weight: 125
x-mbx-used-weight-1m: 125
x-mbx-order-count-10s: 1
x-mbx-order-count-1d: 1
Strict-Transport-Security: max-age=31536000; includeSubdomains
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Content-Security-Policy: default-src 'self'
X-WebKit-CSP: default-src 'self'
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
X-Cache: Miss from cloudfront
Via: 1.1 1e6d2b7781bdded089f18a79a3020a62.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SIN2-P3
X-Amz-Cf-Id: 9QQgCNxmBJ3Key02ap1IKDj995lmpDVG0irp3-c2T8DR_T2EyqKfww==

{"symbol":"PEPEUSDT","orderId":5410202348,"orderListId":-1,"clientOrderId":"MM-0a48228afa3c42ab","transactTime":1788596208102,"price":"0.00000352","origQty":"312945.00","executedQty":"312945.00","origQuoteOrderQty":"0.00000000","cummulativeQuoteQty":"1.10156640","status":"FILLED","timeInForce":"GTC","type":"LIMIT","side":"SELL","workingTime":1788596208102,"fills":[{"price":"0.00000352","qty":"312945.00","commission":"0.00110157","commissionAsset":"USDT","tradeId":560398418}],"selfTradePreventionMode":"EXPIRE_MAKER"}
stream.client_oid: MM-0a48228afa3c42ab, stream.exec_type: NEW
- PLACE LIMIT ORDER - client_oid: MM-0a48228afa3c42ab, status: LIVE, timestamp: 1788596208102
stream NEW1
stream END
stream.client_oid: MM-0a48228afa3c42ab, stream.exec_type: TRADE
stream TRADE
stream FILLED
- FILLED LIMIT ORDER - client_oid: MM-0a48228afa3c42ab, status: FILLED, timestamp: 1788596208102
stream FILLED 1
fills_file_path: data/runs/run_20260905_081614/fills_000002.parquet
stream TRADE 1
stream END
events_file_path: data/runs/run_20260905_081614/events_000002.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 76 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000002.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -354 ms best_rtt: 81 ms
trades_file_path: data/runs/run_20260905_081614/trades_000002.parquet
events_file_path: data/runs/run_20260905_081614/events_000003.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 80 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000003.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 77 ms
events_file_path: data/runs/run_20260905_081614/events_000004.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000004.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 78 ms
events_file_path: data/runs/run_20260905_081614/events_000005.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -354 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 77 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000005.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
events_file_path: data/runs/run_20260905_081614/events_000007.parquetevents_file_path: data/runs/run_20260905_081614/events_000006.parquet

base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 76 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000006.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 78 ms
events_file_path: data/runs/run_20260905_081614/events_000008.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 76 ms
trades_file_path: data/runs/run_20260905_081614/trades_000003.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 76 ms
events_file_path: data/runs/run_20260905_081614/events_000009.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000007.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 76 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -358 ms best_rtt: 76 ms
events_file_path: data/runs/run_20260905_081614/events_000010.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 77 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000008.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 78 ms
events_file_path: events_file_path: data/runs/run_20260905_081614/events_000012.parquet
data/runs/run_20260905_081614/events_000011.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 78 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000009.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 78 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 77 ms
trades_file_path: data/runs/run_20260905_081614/trades_000004.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 79 ms
events_file_path: data/runs/run_20260905_081614/events_000013.parquet
trades_file_path: data/runs/run_20260905_081614/trades_000005.parquet
trades_file_path: data/runs/run_20260905_081614/trades_000006.parquet
trades_file_path: data/runs/run_20260905_081614/trades_000007.parquet
- CANCEL LIMIT ORDER - client_oid: MM-8da57fc5657843b0, status: PENDING_CANCEL, timestamp: 1788596372263
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596372662,"s":"PEPEUSDT","c":"tvvMJAEPrS8grNzhSqUsBv","S":"BUY","o":"LIMIT","f":"GTC","q":"312945.00","p":"0.00000351","P":"0.00000000","F":"0.00","g":-1,"C":"MM-8da57fc5657843b0","x":"CANCELED","X":"CANCELED","r":"NONE","i":5410202117,"l":"0.00","z":"0.00","L":"0.00000000","n":"0","N":null,"T":1788596372660,"t":-1,"I":11553803503,"w":false,"m":false,"M":false,"O":1788596178946,"Z":"0.00000000","Y":"0.00000000","Q":"0.00000000","W":1788596178946,"V":"EXPIRE_MAKER"}}

===== HTTP RESPONSE =====
HTTP/1.1 200 OK
Content-Type: application/json;charset=UTF-8
Content-Length: 412
Connection: keep-alive
Date: Sat, 05 Sep 2026 08:19:32 GMT
Server: nginx
Vary: Accept-Encoding
x-mbx-uuid: a937c45b-8f02-40b7-bbfb-2239de382dae
x-mbx-used-weight: 26
x-mbx-used-weight-1m: 26
Strict-Transport-Security: max-age=31536000; includeSubdomains
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Content-Security-Policy: default-src 'self'
X-WebKit-CSP: default-src 'self'
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
X-Cache: Miss from cloudfront
Via: 1.1 1e6d2b7781bdded089f18a79a3020a62.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SIN2-P3
X-Amz-Cf-Id: KqoVbkLucjFrZlTkEOxNZtPbPBji8MYSiWshAunA_HKivtxMqlguMw==

{"symbol":"PEPEUSDT","origClientOrderId":"MM-8da57fc5657843b0","orderId":5410202117,"orderListId":-1,"clientOrderId":"tvvMJAEPrS8grNzhSqUsBv","transactTime":1788596372660,"price":"0.00000351","origQty":"312945.00","executedQty":"0.00","origQuoteOrderQty":"0.00000000","cummulativeQuoteQty":"0.00000000","status":"CANCELED","timeInForce":"GTC","type":"LIMIT","side":"BUY","selfTradePreventionMode":"EXPIRE_MAKER"}
stream.client_oid: MM-8da57fc5657843b0, stream.exec_type: CANCELED
- CANCEL LIMIT ORDER - client_oid: MM-8da57fc5657843b0, status: CANCELED, timestamp: 1788596372660
stream END
- PLACE LIMIT ORDER - client_oid: MM-d02bda20279f4a0a, status: PENDING_NEW, timestamp: 1788596372355
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596372751,"s":"PEPEUSDT","c":"MM-d02bda20279f4a0a","S":"BUY","o":"LIMIT","f":"GTC","q":"830286.00","p":"0.00000350","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"NEW","X":"NEW","r":"NONE","i":5410203142,"l":"0.00","z":"0.00","L":"0.00000000","n":"0","N":null,"T":1788596372750,"t":-1,"I":11553803532,"w":true,"m":false,"M":false,"O":1788596372750,"Z":"0.00000000","Y":"0.00000000","Q":"0.00000000","W":1788596372750,"V":"EXPIRE_MAKER"}}

===== HTTP RESPONSE =====
HTTP/1.1 200 OK
Content-Type: application/json;charset=UTF-8
Content-Length: 401
Connection: keep-alive
Date: Sat, 05 Sep 2026 08:19:32 GMT
Server: nginx
Vary: Accept-Encoding
x-mbx-uuid: cdadb698-f0da-4eb2-8b89-0c5e9cd9ac9a
x-mbx-used-weight: 26
x-mbx-used-weight-1m: 26
x-mbx-order-count-10s: 1
x-mbx-order-count-1d: 1
Strict-Transport-Security: max-age=31536000; includeSubdomains
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Content-Security-Policy: default-src 'self'
X-WebKit-CSP: default-src 'self'
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
X-Cache: Miss from cloudfront
Via: 1.1 1e6d2b7781bdded089f18a79a3020a62.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SIN2-P3
X-Amz-Cf-Id: tf-9PyuF9-vqCBqYMCPHn_NbGeA7Jp7WNquRsCVMJdwe2nsfUNQ6Qg==

{"symbol":"PEPEUSDT","orderId":5410203142,"orderListId":-1,"clientOrderId":"MM-d02bda20279f4a0a","transactTime":1788596372750,"price":"0.00000350","origQty":"830286.00","executedQty":"0.00","origQuoteOrderQty":"0.00000000","cummulativeQuoteQty":"0.00000000","status":"NEW","timeInForce":"GTC","type":"LIMIT","side":"BUY","workingTime":1788596372750,"fills":[],"selfTradePreventionMode":"EXPIRE_MAKER"}
stream.client_oid: MM-d02bda20279f4a0a, stream.exec_type: NEW
- PLACE LIMIT ORDER - client_oid: MM-d02bda20279f4a0a, status: LIVE, timestamp: 1788596372750
quotes_file_path: data/runs/run_20260905_081614/quotes_000001.parquet
stream NEW1
stream END
events_file_path: data/runs/run_20260905_081614/events_000014.parquet
- CANCEL LIMIT ORDER - client_oid: MM-d02bda20279f4a0a, status: PENDING_CANCEL, timestamp: 1788596373029
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596373432,"s":"PEPEUSDT","c":"cUe82AAb59zj5fITbnlPBM","S":"BUY","o":"LIMIT","f":"GTC","q":"830286.00","p":"0.00000350","P":"0.00000000","F":"0.00","g":-1,"C":"MM-d02bda20279f4a0a","x":"CANCELED","X":"CANCELED","r":"NONE","i":5410203142,"l":"0.00","z":"0.00","L":"0.00000000","n":"0","N":null,"T":1788596373431,"t":-1,"I":11553803672,"w":false,"m":false,"M":false,"O":1788596372750,"Z":"0.00000000","Y":"0.00000000","Q":"0.00000000","W":1788596372750,"V":"EXPIRE_MAKER"}}

===== HTTP RESPONSE =====
HTTP/1.1 200 OK
Content-Type: application/json;charset=UTF-8
Content-Length: 412
Connection: keep-alive
Date: Sat, 05 Sep 2026 08:19:33 GMT
Server: nginx
Vary: Accept-Encoding
x-mbx-uuid: 877ff971-931a-4587-85ea-9aecca86e76a
x-mbx-used-weight: 26
x-mbx-used-weight-1m: 26
Strict-Transport-Security: max-age=31536000; includeSubdomains
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Content-Security-Policy: default-src 'self'
X-WebKit-CSP: default-src 'self'
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
X-Cache: Miss from cloudfront
Via: 1.1 1e6d2b7781bdded089f18a79a3020a62.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SIN2-P3
X-Amz-Cf-Id: Q1GMMdZIZTddidESaD9jItezJxKFgKKznf9JrKzL3DPbBNe7LQz9xg==

{"symbol":"PEPEUSDT","origClientOrderId":"MM-d02bda20279f4a0a","orderId":5410203142,"orderListId":-1,"clientOrderId":"cUe82AAb59zj5fITbnlPBM","transactTime":1788596373431,"price":"0.00000350","origQty":"830286.00","executedQty":"0.00","origQuoteOrderQty":"0.00000000","cummulativeQuoteQty":"0.00000000","status":"CANCELED","timeInForce":"GTC","type":"LIMIT","side":"BUY","selfTradePreventionMode":"EXPIRE_MAKER"}
stream.client_oid: MM-d02bda20279f4a0a, stream.exec_type: CANCELED
- CANCEL LIMIT ORDER - client_oid: MM-d02bda20279f4a0a, status: CANCELED, timestamp: 1788596373431
stream END
- PLACE LIMIT ORDER - client_oid: MM-e70cce436d284769, status: PENDING_NEW, timestamp: 1788596373117
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596373514,"s":"PEPEUSDT","c":"MM-e70cce436d284769","S":"BUY","o":"LIMIT","f":"GTC","q":"830286.00","p":"0.00000351","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"NEW","X":"NEW","r":"NONE","i":5410203219,"l":"0.00","z":"0.00","L":"0.00000000","n":"0","N":null,"T":1788596373514,"t":-1,"I":11553803679,"w":true,"m":false,"M":false,"O":1788596373514,"Z":"0.00000000","Y":"0.00000000","Q":"0.00000000","W":1788596373514,"V":"EXPIRE_MAKER"}}

===== HTTP RESPONSE =====
HTTP/1.1 200 OK
Content-Type: application/json;charset=UTF-8
Content-Length: 401
Connection: keep-alive
Date: Sat, 05 Sep 2026 08:19:33 GMT
Server: nginx
Vary: Accept-Encoding
x-mbx-uuid: 5ef313c7-5ea3-42ac-984d-a50296208d6d
x-mbx-used-weight: 26
x-mbx-used-weight-1m: 26
x-mbx-order-count-10s: 2
x-mbx-order-count-1d: 2
Strict-Transport-Security: max-age=31536000; includeSubdomains
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
X-Content-Security-Policy: default-src 'self'
X-WebKit-CSP: default-src 'self'
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, OPTIONS
X-Cache: Miss from cloudfront
Via: 1.1 1e6d2b7781bdded089f18a79a3020a62.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SIN2-P3
X-Amz-Cf-Id: bsBgzAFhI1V8VzdkLN3OeoOHO2ZS2AUIqQoPi_hmkmDdiZdKsdbVzw==

{"symbol":"PEPEUSDT","orderId":5410203219,"orderListId":-1,"clientOrderId":"MM-e70cce436d284769","transactTime":1788596373514,"price":"0.00000351","origQty":"830286.00","executedQty":"0.00","origQuoteOrderQty":"0.00000000","cummulativeQuoteQty":"0.00000000","status":"NEW","timeInForce":"GTC","type":"LIMIT","side":"BUY","workingTime":1788596373514,"fills":[],"selfTradePreventionMode":"EXPIRE_MAKER"}
stream.client_oid: MM-e70cce436d284769, stream.exec_type: NEW
- PLACE LIMIT ORDER - client_oid: MM-e70cce436d284769, status: LIVE, timestamp: 1788596373514
stream NEW1
stream END
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000010.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -354 ms best_rtt: 75 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
events_file_path: data/runs/run_20260905_081614/events_000015.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000011.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -354 ms best_rtt: 78 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 77 ms
events_file_path: data/runs/run_20260905_081614/events_000016.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000012.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 79 ms
events_file_path: data/runs/run_20260905_081614/events_000017.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 80 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000013.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -354 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -358 ms best_rtt: 78 ms
events_file_path: data/runs/run_20260905_081614/events_000018.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000014.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 78 ms
events_file_path: data/runs/run_20260905_081614/events_000019.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000015.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 80 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -359 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 81 ms
events_file_path: data/runs/run_20260905_081614/events_000020.parquet
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000016.parquet
trades_file_path: data/runs/run_20260905_081614/trades_000008.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -359 ms best_rtt: 76 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 76 ms
events_file_path: data/runs/run_20260905_081614/events_000021.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 78 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000017.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 76 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
events_file_path: data/runs/run_20260905_081614/events_000022.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 79 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000018.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 77 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 77 ms
events_file_path: data/runs/run_20260905_081614/events_000023.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 74 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 77 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000019.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 79 ms
events_file_path: data/runs/run_20260905_081614/events_000024.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 78 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -358 ms best_rtt: 78 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000020.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 77 ms
events_file_path: data/runs/run_20260905_081614/events_000025.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 79 ms
trades_file_path: data/runs/run_20260905_081614/trades_000009.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 76 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000021.parquet
events_file_path: data/runs/run_20260905_081614/events_000026.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -355 ms best_rtt: 78 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -354 ms best_rtt: 77 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000022.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 80 ms
events_file_path: data/runs/run_20260905_081614/events_000027.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -357 ms best_rtt: 81 ms
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 75 ms
snapshots_file_path: data/runs/run_20260905_081614/snapshots_000023.parquet
events_file_path: data/runs/run_20260905_081614/events_000028.parquet
base_url: api.binance.com endpoint: /api/v3/time [CLOCK] best_offset: -356 ms best_rtt: 75 ms
{"subscriptionId":0,"event":{"e":"executionReport","E":1788596622383,"s":"PEPEUSDT","c":"MM-e70cce436d284769","S":"BUY","o":"LIMIT","f":"GTC","q":"830286.00","p":"0.00000351","P":"0.00000000","F":"0.00","g":-1,"C":"","x":"TRADE","X":"FILLED","r":"NONE","i":5410203219,"l":"830286.00","z":"830286.00","L":"0.00000351","n":"830.29","N":"PEPE","T":1788596622382,"t":560398597,"I":11553805693,"w":false,"m":true,"M":true,"O":1788596373514,"Z":"2.91430386","Y":"2.91430386","Q":"0.00000000","W":1788596373514,"V":"EXPIRE_MAKER"}}
trades_file_path: data/runs/run_20260905_081614/trades_000010.parquet
events_file_path: data/runs/run_20260905_081614/events_000029.parquet
stream.client_oid: MM-e70cce436d284769, stream.exec_type: TRADE
stream TRADE
stream FILLED
- FILLED LIMIT ORDER - client_oid: MM-e70cce436d284769, status: FILLED, timestamp: 1788596622382
stream FILLED 1
fills_file_path: data/runs/run_20260905_081614/fills_000003.parquet
stream TRADE 1
stream END
- PLACE LIMIT ORDER - client_oid: MM-4b8cbb2adf8948be, status: PENDING_NEW, timestamp: 1788596622085
PS D:\OneDrive\Trading\Market Making>

In [18]:
snapshots = pd.read_parquet(os.path.join(folder_path, "snapshots.parquet"))
snapshots["micro_signal"] = (
    snapshots["microprice"] - snapshots["mid"]
) / snapshots["mid"]

snapshot_cols = [
    "ts",
    "symbol",
    "mid",
    "microprice",
    "micro_signal",
    "microprice_error",
    "spread",
    "volatility",
    "order_imbalance",
    "trade_imbalance",
    "quote_churn",
    "inventory",
    "regime",
]

snap = snapshots[snapshot_cols].copy()
snap = snap.rename(columns={"ts": "snapshot_ts"})

fills2 = fills.copy()
fills2 = fills2.rename(columns={"ts": "fill_ts"})

snap = snap.sort_values(["symbol", "snapshot_ts"])
fills2 = fills2.sort_values(["symbol", "fill_ts"])

fills_enriched = pd.merge_asof(
    fills2,
    snap,
    left_on="fill_ts",
    right_on="snapshot_ts",
    by="symbol",
    direction="backward",
)

fills_enriched["snapshot_age_ms"] = (
    fills_enriched["fill_ts"] -
    fills_enriched["snapshot_ts"]
)

fills_enriched

,fill_ts,exchange_latency,time_to_fill,symbol,side,price,price_tick,qty,cum_qty,is_maker,...,micro_signal,microprice_error_y,spread_y,volatility_y,order_imbalance_y,trade_imbalance_y,quote_churn,inventory_y,regime_y,snapshot_age_ms
0,1787985394060,0,92178,PEPEUSDT,SELL,0.000004,363,2.0,2.0,True,...,-0.000085,3.098560e-10,1.000000e-08,0.001380,-0.061971,-0.143542,0.0,0.0,no_regime_model,816
1,1787985437496,0,43172,PEPEUSDT,SELL,0.000004,363,1.0,1.0,True,...,0.000878,-3.182027e-09,1.000000e-08,0.001380,0.636405,-0.571274,0.0,-2.0,no_regime_model,52
2,1787985604374,0,166791,PEPEUSDT,SELL,0.000004,363,1.0,1.0,True,...,-0.000041,1.480424e-10,1.000000e-08,0.001380,-0.029608,-0.495890,0.0,-3.0,no_regime_model,930
3,1787985730728,0,126247,PEPEUSDT,SELL,0.000004,363,1.0,1.0,True,...,0.000957,-3.468017e-09,1.000000e-08,0.001380,0.693603,0.603717,0.0,-4.0,no_regime_model,84
4,1787985836923,0,535040,PEPEUSDT,BUY,0.000004,362,2.0,2.0,True,...,0.000839,-3.041278e-09,1.000000e-08,0.001419,0.608256,-0.891040,0.0,-5.0,no_regime_model,379
5,1787985905607,0,68624,PEPEUSDT,SELL,0.000004,363,1.0,1.0,True,...,0.000938,-3.401019e-09,1.000000e-08,0.001380,0.680204,0.278850,0.0,-3.0,no_regime_model,363
6,1787985934804,0,29114,PEPEUSDT,SELL,0.000004,363,1.0,1.0,True,...,0.001150,-4.169676e-09,1.000000e-08,0.001380,0.833935,0.423080,0.0,-4.0,no_regime_model,60
7,1787986522292,0,685311,PEPEUSDT,BUY,0.000004,362,3.0,3.0,True,...,0.001061,-3.847079e-09,1.000000e-08,0.001380,0.769416,0.912129,0.0,-5.0,no_regime_model,1048
8,1787986728867,0,206478,PEPEUSDT,SELL,0.000004,363,1.0,1.0,True,...,0.000861,-3.121311e-09,1.000000e-08,0.001380,0.624262,-0.985163,0.0,-2.0,no_regime_model,823
9,1787986824614,0,95625,PEPEUSDT,SELL,0.000004,363,1.0,1.0,True,...,0.000934,-3.387337e-09,1.000000e-08,0.001380,0.677467,0.436747,0.0,-3.0,no_regime_model,370


In [14]:
fills_enriched["snapshot_age_ms"].describe()

count      43.000000
mean      562.395349
std       859.592765
min         1.000000
25%        59.500000
50%       229.000000
75%       819.500000
max      4750.000000
Name: snapshot_age_ms, dtype: float64

In [ ]:
fills_enriched["signed_micro_signal"] = (
    fills_enriched["micro_signal"]
    * fills_enriched["side"].map({
        "BUY": 1,
        "SELL": -1
    })
)

# fills_enriched["signed_micro_signal"] = (
#     fills_enriched["micro_signal"]
#     * fills_enriched["fill_sign"]
# )

fills_enriched["signal_bucket"] = pd.qcut(
    fills_enriched["signed_micro_signal"],
    10,
    labels=False,
    duplicates="drop"
)

result = (
    fills_enriched
    .groupby("signal_bucket")
    .agg(
        fill_count=("signed_micro_signal", "size"),
        signal_mean=("signed_micro_signal", "mean"),
        mo_100ms=("signed_markout_100ms", "mean"),
        mo_500ms=("signed_markout_500ms", "mean"),
        mo_1000ms=("signed_markout_1000ms", "mean"),
        mo_5000ms=("signed_markout_5000ms", "mean"),
    )
)

print(result)

               fill_count  signal_mean      mo_100ms      mo_500ms  \
signal_bucket                                                        
0                       5    -0.001250 -3.000000e-09 -3.000000e-09   
1                       4    -0.001123  0.000000e+00  0.000000e+00   
2                       4    -0.000979  2.500000e-09  2.500000e-09   
3                       4    -0.000934  2.500000e-09  2.500000e-09   
4                       5    -0.000791  3.000000e-09  3.000000e-09   
5                       4    -0.000268  5.000000e-09  5.000000e-09   
6                       4    -0.000183  0.000000e+00  0.000000e+00   
7                       4    -0.000009  5.000000e-09  5.000000e-09   
8                       4     0.000087  5.000000e-09  5.000000e-09   
9                       5     0.000995  5.000000e-09  5.000000e-09   

                  mo_1000ms     mo_5000ms  
signal_bucket                              
0             -3.000000e-09 -1.000000e-09  
1              0.000000e+00

In [24]:
import statsmodels.api as sm

x = fills_enriched["signed_micro_signal"]
y = fills_enriched["signed_markout_5000ms"]

X = sm.add_constant(x)

model = sm.OLS(y, X).fit()

print(model.summary())

                              OLS Regression Results                             
Dep. Variable:     signed_markout_5000ms   R-squared:                       0.122
Model:                               OLS   Adj. R-squared:                  0.100
Method:                    Least Squares   F-statistic:                     5.671
Date:                   Tue, 01 Sep 2026   Prob (F-statistic):             0.0220
Time:                           08:39:25   Log-Likelihood:                 768.69
No. Observations:                     43   AIC:                            -1533.
Df Residuals:                         41   BIC:                            -1530.
Df Model:                              1                                         
Covariance Type:               nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
cons

In [21]:
result = (
    fills_enriched
    .groupby("signal_bucket")
    .agg(
        n=("signed_micro_signal", "size"),
        signal=("signed_micro_signal", "mean"),
        mo100=("signed_markout_100ms", "mean"),
        mo500=("signed_markout_500ms", "mean"),
        mo1s=("signed_markout_1000ms", "mean"),
        mo5s=("signed_markout_5000ms", "mean"),
    )
)

g = fills_enriched.groupby("signal_bucket")

result = g["signed_markout_5000ms"].agg(
    n="count",
    mean="mean",
    std="std"
)

result["se"] = result["std"] / np.sqrt(result["n"])
result["ci95"] = 1.96 * result["se"]


In [22]:
result

,n,mean,std,se,ci95
signal_bucket,,,,,
0,5,-1.000000e-09,5.477226e-09,2.449490e-09,4.801000e-09
1,4,0.000000e+00,5.773503e-09,2.886751e-09,5.658033e-09
2,4,5.000000e-09,0.000000e+00,0.000000e+00,0.000000e+00
3,4,5.000000e-09,0.000000e+00,0.000000e+00,0.000000e+00
4,5,3.000000e-09,4.472136e-09,2.000000e-09,3.920000e-09
5,4,2.500000e-09,5.000000e-09,2.500000e-09,4.900000e-09
6,4,0.000000e+00,5.773503e-09,2.886751e-09,5.658033e-09
7,4,5.000000e-09,0.000000e+00,0.000000e+00,0.000000e+00
8,4,5.000000e-09,0.000000e+00,0.000000e+00,0.000000e+00


In [17]:
for i, j in fills_enriched.iloc[0].items():
    print(i, j)

fill_ts 1787985394060
exchange_latency 0
time_to_fill 92178
symbol PEPEUSDT
side SELL
price 3.63e-06
price_tick 363
qty 2.0
cum_qty 2.0
is_maker True
fill_sign -1
fill_type ASK_LIFT
fill_status FULL
queue_ahead_at_join 45717907304.0
mid_at_fill 3.625e-06
microprice_at_fill 3.628065936116282e-06
microprice_dev_at_fill 3.0659361162820746e-09
microprice_error_at_fill -3.0659361162820746e-09
spread_at_fill 9.999999999999904e-09
best_bid_at_fill 3.62e-06
best_ask_at_fill 3.63e-06
volatility_at_fill 0.0013802624675619511
volatility_at_fill_bps 13.80262467561951
mid_x 3.625e-06
microprice_x 3.6247074475239223e-06
microprice_dev -2.925524760777309e-10
microprice_error_x 2.925524760777309e-10
spread_x 9.999999999999904e-09
best_bid 3.62e-06
best_ask 3.63e-06
best_bid_tick 362
best_ask_tick 363
order_imbalance_x -0.05851049521551501
trade_imbalance_x 0.0
volatility_x 0.0
volatility_bps 0.0
queue_ahead_bid 0.0
queue_ahead_ask 0.0
inventory_x -2.0
fair -0.01169847433565548
skew -0.0
struct_delta 0

In [2]:
"""
model = predict adverse selection

                ┌──────────────┐
market ───────► │ regime model │ ───► policy params
                └──────┬───────┘
                       │
                       ▼
        ┌──────────────────────────┐
        │ alpha / fair value stack │ ───► center
        └──────────┬───────────────┘
                   │
                   ▼
        ┌──────────────────────────┐
        │ toxicity model           │ ───► risk modifiers
        └──────────┬───────────────┘
                   │
                   ▼
             execution engine

So your best toxicity model is actually:

E[future markout loss | fill now]

That is the true label your XGBoost model should learn.

Not classification. Not heuristic imbalance.
"""

df = fills

horizons = [100, 500, 1000, 5000]

feature_cols = [
    "microprice_dev",
    "order_imbalance",
    "trade_imbalance",
    "volatility",
    "spread",
    "queue_ahead_bid",
    "queue_ahead_ask",
]

df = df.dropna()

split = int(len(df) * 0.8)
train = df.iloc[:split]
test = df.iloc[split:]

X_train = train[feature_cols].to_numpy(dtype=np.float32)
X_test = test[feature_cols].to_numpy(dtype=np.float32)

In [5]:
def ic(pred, y):
    return np.corrcoef(pred, y)[0, 1]

def rank_ic(pred, y):
    return spearmanr(pred, y).statistic

results = {}
models = {}

for h in horizons:

    y_train = train[f"signed_markout_{h}ms"]
    y_test = test[f"signed_markout_{h}ms"]

    model = xgb.XGBRegressor(
        n_estimators=300,
        max_depth=4,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=42
    )

    model.fit(X_train, y_train)

    train_pred = model.predict(X_train)
    tox_std = np.std(train_pred)
    tox_mean_abs = np.mean(np.abs(train_pred))

    pred = model.predict(X_test)
    true = y_test.values

    # -------------------------
    # signal quality
    # -------------------------
    residual_ic = ic(pred, true)
    residual_rank_ic = rank_ic(pred, true)

    # -------------------------
    # direction accuracy
    # -------------------------
    hit_rate = (np.sign(pred) == np.sign(true)).mean()

    # -------------------------
    # true economic interpretation
    # -------------------------
    pnl_proxy = np.mean(pred * true)
    pnl_std = np.std(pred * true) + 1e-9
    sharpe_proxy = pnl_proxy / pnl_std

    models[h] = {
        "model": model,
        "tox_std": tox_std,
        "tox_mean_abs": tox_mean_abs
    }
    results[h] = {
        "Residual_IC": residual_ic,
        "Residual_Rank_IC": residual_rank_ic,
        "HitRate": hit_rate,
        "PnLProxy": pnl_proxy,
        "SharpeProxy": sharpe_proxy,
    }

results_df = pd.DataFrame(results)
results_df

c:\Users\brian\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\numpy\lib\_function_base_impl.py:3036: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
c:\Users\brian\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\numpy\lib\_function_base_impl.py:3037: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
C:\Users\brian\AppData\Local\Temp\ipykernel_23180\1801115838.py:5: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  return spearmanr(pred, y).statistic


,100,500,1000,5000
Residual_IC,NaN,NaN,NaN,NaN
Residual_Rank_IC,NaN,NaN,NaN,NaN
HitRate,7.777778e-01,7.777778e-01,7.777778e-01,6.666667e-01
PnLProxy,6.649031e-18,6.649031e-18,5.880236e-18,9.117489e-18
SharpeProxy,6.649031e-09,6.649031e-09,5.880236e-09,9.117489e-09


In [11]:
"""
SharpeProxy increasing with horizon

This is actually the key signal:

100ms → weak economic signal
500-1000ms → stronger signal

This tells you:

toxicity is not instantaneous micro-noise, it is short-horizon information flow

That's exactly what informed order flow looks like.

3. The important conceptual insight

You are NOT building:

a price predictor

You ARE building:

a liquidity filter

This model answers:

“Should I provide liquidity here or not?”

"""

"\nSharpeProxy increasing with horizon\n\nThis is actually the key signal:\n\n100ms → weak economic signal\n500-1000ms → stronger signal\n\nThis tells you:\n\ntoxicity is not instantaneous micro-noise, it is short-horizon information flow\n\nThat's exactly what informed order flow looks like.\n\n3. The important conceptual insight\n\nYou are NOT building:\n\na price predictor\n\nYou ARE building:\n\na liquidity filter\n\nThis model answers:\n\n“Should I provide liquidity here or not?”\n\n"

In [ ]:
# artifact = {
#     "model": models[100]["model"],
#     "feature_cols": feature_cols,
#     "target": "markout_100ms",
#     "horizon_ms": 100,
# }

# joblib.dump(artifact, "data/toxicity_model.pkl")

['data/toxicity_model.pkl']

In [13]:
artifact = {
    "name": "toxicity_model",
    "model_file": "toxicity_model_xgb.json",
    "feature_cols": feature_cols,
    "feature_dim": len(feature_cols),
    "target": "markout_100ms",
    "horizon_ms": 100
}

models[100]["model"].save_model("data/toxicity_model_xgb.json")

with open("data/toxicity_model.json", "w") as f:
    json.dump(artifact, f, indent=4)

In [14]:
def export_xgb(model_name, target, horizon_ms):
    model = models[horizon_ms]["model"]
    model.save_model(f"data/{model_name}_xgb.json")

    artifact = {
        "model_name": f"{model_name}",
        "target": target,
        "model_file": f"data/{model_name}_xgb.json",
        "horizon_ms": horizon_ms,
        "feature_cols": feature_cols,
        "feature_dim": len(feature_cols)
    }

    with open(f"data/{model_name}.json", "w") as f:
        json.dump(artifact, f, indent=4)
    
    print(f"['data/{model_name}.json']")
    print(f"['data/{model_name}_xgb.json']")

export_xgb(model_name="toxicity_model_2", target="markout_100ms", horizon_ms=100)

['data/toxicity_model_2.json']
['data/toxicity_model_2_xgb.json']
